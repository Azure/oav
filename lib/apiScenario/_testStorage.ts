import { inversifyGetInstance } from "../inversifyUtils";
import "reflect-metadata";
import { PostmanCollectionGenerator } from "./postmanCollectionGenerator";

// oav generate-postmanCollection testScenario --env
const main = async () => {
  const env = {
    subscriptionId: "db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    location: "eastasia",
    scope: "subscriptions/db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    resourceId: "mysourceid",
    tenantId: "<your tenantId>",
    client_id: "<your client_id>",
    client_secret: "<your client secret>",
    resourceGroupName: "ruowan-test",
    deploymentName: "createVMSS",
  };

  const postmanCollectionGenerator = inversifyGetInstance(PostmanCollectionGenerator, {
    name: "storage",
    fileRoot: "/home/ruowan/work/azure-rest-api-specs/specification/storage/resource-manager",
    swaggerFilePaths: [
      "Microsoft.Storage/stable/2019-06-01/storage.json",
      "Microsoft.Storage/stable/2019-06-01/blob.json",
    ],
    testDef: "Microsoft.Storage/stable/2019-06-01/test-scenairos/createStorageAccount.yaml",
    env: env,
    outputFolder: "./fixtures",
    generateCollection: true,
    runCollection: false,
  });

  try {
    await postmanCollectionGenerator.GenerateCollection();
  } catch (e) {
    console.log(e.message, e.stack);
  }
};

console.time("TestLoad");
console.log("Start");

main().finally(() => {
  console.timeEnd("TestLoad");
});
