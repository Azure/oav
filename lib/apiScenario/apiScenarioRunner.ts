import { HttpMethods } from "@azure/core-http";
import { JsonLoader } from "../swagger/jsonLoader";
import { getLazyBuilder } from "../util/lazyBuilder";
import { getRandomString } from "../util/utils";
import {
  ArmTemplate,
  ScenarioDefinition,
  Scenario,
  Step,
  StepArmTemplate,
  StepRestCall,
} from "./apiScenarioTypes";
import { VariableEnv } from "./variableEnv";

export interface ApiScenarioRunnerOption {
  env: VariableEnv;
  client: ApiScenarioRunnerClient;
  jsonLoader: JsonLoader;
  resolveVariables?: boolean;
}

export interface ArmDeploymentTracking {
  deploymentName: string;
  step: StepArmTemplate;
  details: {
    scope: "ResourceGroup";
    subscriptionId: string;
    resourceGroupName: string;
  };
}

interface ScopeTracking {
  provisioned?: boolean;
  scope: ScenarioDefinition["scope"];
  prepareSteps: Step[];
  cleanUpSteps: Step[];
  env: VariableEnv;
  armDeployments: ArmDeploymentTracking[];
}

export interface ApiScenarioClientRequest {
  method: HttpMethods;
  path: string;
  pathVariables?: { [variableName: string]: string };
  headers: { [headerName: string]: string };
  query: { [key: string]: string };
  body?: any;
}

export interface ApiScenarioRunnerClient {
  createResourceGroup(
    subscriptionId: string,
    resourceGroupName: string,
    location: string
  ): Promise<void>;

  deleteResourceGroup(subscriptionId: string, resourceGroupName: string): Promise<void>;

  sendRestCallRequest(
    request: ApiScenarioClientRequest,
    step: StepRestCall,
    env: VariableEnv
  ): Promise<void>;

  sendArmTemplateDeployment(
    armTemplate: ArmTemplate,
    armDeployment: ArmDeploymentTracking,
    step: StepArmTemplate,
    env: VariableEnv
  ): Promise<void>;
}

export class ApiScenarioRunner {
  private jsonLoader: JsonLoader;
  private client: ApiScenarioRunnerClient;
  private env: VariableEnv;
  private scopeTracking: { [scopeName: string]: ScopeTracking };
  private scenarioScopeTracking: Map<Scenario, ScopeTracking>;
  private resolveVariables: boolean;

  private provisionScope = getLazyBuilder("provisioned", async (scope: ScopeTracking) => {
    if (scope.scope !== "ResourceGroup") {
      throw new Error(`Scope is not supported yet: ${scope.scope}`);
    }

    if (scope.env.get("resourceGroupName") === undefined) {
      const resourceGroupPrefix = scope.env.getString("resourceGroupPrefix") ?? "apiTest-";
      scope.env.set("resourceGroupName", {
        type: "string",
        value: resourceGroupPrefix + getRandomString(),
      });
    }

    await this.client.createResourceGroup(
      scope.env.getRequiredString("subscriptionId"),
      scope.env.getRequiredString("resourceGroupName"),
      scope.env.getRequiredString("location")
    );

    for (const step of scope.prepareSteps) {
      await this.executeStep(step, scope.env, scope);
    }

    return true;
  });

  public constructor(opts: ApiScenarioRunnerOption) {
    this.env = opts.env;
    this.client = opts.client;
    this.jsonLoader = opts.jsonLoader;
    this.resolveVariables = opts.resolveVariables ?? true;
    this.scopeTracking = {};
    this.scenarioScopeTracking = new Map();
  }

  public async prepareScope(scenario: Scenario): Promise<ScopeTracking> {
    let scope = this.scenarioScopeTracking.get(scenario);
    if (scope === undefined) {
      const scopeName = scenario.shareScope ? "_defaultScope" : `_randomScope_${getRandomString()}`;

      scope = this.scopeTracking[scopeName];
      if (scope === undefined) {
        const scenarioDef = scenario._scenarioDef;
        const env = new VariableEnv(this.env);
        env.setBatch(scenarioDef.variables);
        scope = {
          scope: scenarioDef.scope,
          prepareSteps: scenarioDef.prepareSteps,
          cleanUpSteps: scenarioDef.cleanUpSteps,
          env,
          armDeployments: [],
        };
        this.scopeTracking[scopeName] = scope;
      }

      this.scenarioScopeTracking.set(scenario, scope);
    }

    await this.provisionScope(scope);
    return scope;
  }

  public scopePreparedExternal(
    scopeInput: Pick<ScopeTracking, "env" | "armDeployments" | "provisioned">,
    info: {
      scenario?: Scenario;
      scopeName?: string;
      scenarioDef?: ScenarioDefinition;
    }
  ): void {
    const { scenario, scopeName } = info;
    const scenarioDef = scenario?._scenarioDef ?? info.scenarioDef;

    if (scenarioDef === undefined) {
      throw new Error("Either Scenario or ScenarioDef must be provided.");
    }

    const scope = {
      scope: scenarioDef.scope,
      prepareSteps: scenarioDef.prepareSteps,
      cleanUpSteps: scenarioDef.cleanUpSteps,
      ...scopeInput,
    };
    if (scopeName !== undefined) {
      if (this.scopeTracking[scopeName] !== undefined) {
        throw new Error(`Scope already created: ${scopeName}`);
      }
      this.scopeTracking[scopeName] = scope;
    }

    if (scenario !== undefined) {
      if (this.scenarioScopeTracking.get(scenario) !== undefined) {
        throw new Error(
          `Scope already created for scenario: ${scenario.description} , scopeName: ${scopeName}`
        );
      }
      this.scenarioScopeTracking.set(scenario, scope);
    }
  }

  public async executeScenario(scenario: Scenario) {
    const scope = await this.prepareScope(scenario);
    const env = new VariableEnv(scope.env);
    env.setBatch(scenario.variables);

    for (const step of scenario.steps) {
      await this.executeStep(step, env, scope);
    }
  }

  public async executeStep(step: Step, env: VariableEnv, scope: ScopeTracking) {
    const stepEnv = new VariableEnv(env);
    stepEnv.setBatch(step.variables);
    // stepEnv.setDefault(step.defaultValues);

    if (this.resolveVariables) {
      stepEnv.resolve();
    }

    try {
      switch (step.type) {
        case "restCall":
          await this.executeRestCallStep(step, stepEnv);
          break;
        case "armTemplateDeployment":
          await this.executeArmTemplateStep(step, stepEnv, scope);
          break;
      }
    } catch (error) {
      throw new Error(
        `Failed to execute step ${step.step}: ${(error as any).message} \n${error.stack}`
      );
    }
  }

  public async cleanAllScope() {
    for (const scope of Object.values(this.scopeTracking)) {
      for (const step of scope.cleanUpSteps) {
        await this.executeStep(step, scope.env, scope);
      }
      const subscriptionId = scope.env.getRequiredString("subscriptionId");
      const resourceGroupName = scope.env.getRequiredString("resourceGroupName");
      await this.client.deleteResourceGroup(subscriptionId, resourceGroupName);
    }
  }

  private async executeRestCallStep(step: StepRestCall, env: VariableEnv) {
    let req: ApiScenarioClientRequest = {
      method: step.operation._method.toUpperCase() as HttpMethods,
      path: step.operation._path._pathTemplate.replace(/{([a-z0-9_]+)}/gi, (_, p1) => `$(${p1})`),
      pathVariables: {},
      headers: {},
      query: {},
    };

    for (const p of step.operation.parameters ?? []) {
      const param = this.jsonLoader.resolveRefObj(p);

      const paramVal = step.parameters[param.name];
      if (paramVal === undefined && env.get(param.name) === undefined) {
        if (param.required) {
          throw new Error(`Parameter value for "${param.name}" is not found in step: ${step.step}`);
        } else {
          continue;
        }
      }

      switch (param.in) {
        case "path":
          req.pathVariables![param.name] = paramVal;
          break;
        case "query":
          req.query[param.name] = paramVal;
          break;
        case "header":
          req.headers[param.name] = paramVal;
          break;
        case "body":
          req.body = paramVal;
          break;
        default:
          throw new Error(`Parameter "in" not supported: ${param.in}`);
      }
    }

    if (this.resolveVariables) {
      req = env.resolveObjectValues(req);
    }

    await this.client.sendRestCallRequest(req, step, env);
  }

  private async executeArmTemplateStep(
    step: StepArmTemplate,
    env: VariableEnv,
    scope: ScopeTracking
  ) {
    const subscriptionId = env.getRequiredString("subscriptionId");
    const resourceGroupName = env.getRequiredString("resourceGroupName");

    const armDeployment: ArmDeploymentTracking = {
      deploymentName: `${resourceGroupName}-deploy-${getRandomString()}`,
      step,
      details: {
        scope: scope.scope,
        subscriptionId,
        resourceGroupName,
      },
    };
    scope.armDeployments.push(armDeployment);

    if (this.resolveVariables) {
      step.armTemplatePayload = env.resolveObjectValues(step.armTemplatePayload);
    }

    await this.client.sendArmTemplateDeployment(step.armTemplatePayload, armDeployment, step, env);
  }
}
