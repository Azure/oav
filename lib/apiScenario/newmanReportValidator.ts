import * as path from "path";
import { findReadMe } from "@azure/openapi-markdown";
import { inject, injectable } from "inversify";
import { TYPES } from "../inversifyUtils";
import {
  LiveValidator,
  RequestResponseLiveValidationResult,
  RequestResponsePair,
} from "../liveValidation/liveValidator";
import { LiveRequest, LiveResponse } from "../liveValidation/operationValidator";
import { FileLoader } from "../swagger/fileLoader";
import { setDefaultOpts } from "../swagger/loader";
import { SwaggerExample } from "../swagger/swaggerTypes";
import { SeverityString } from "../util/severity";
import { getApiVersionFromFilePath, getProviderFromFilePath } from "../util/utils";
import { ApiScenarioLoaderOption } from "./apiScenarioLoader";
import { NewmanExecution, NewmanReport, Scenario, Step, StepRestCall } from "./apiScenarioTypes";
import { DataMasker } from "./dataMasker";
import { getJsonPatchDiff } from "./diffUtils";
import { JUnitReporter } from "./junitReport";
import { generateMarkdownReport } from "./markdownReport";
import { SwaggerAnalyzer } from "./swaggerAnalyzer";
import { VariableEnv } from "./variableEnv";

interface GeneratedExample {
  step: string;
  operationId: string;
  example: SwaggerExample;
}

export interface ApiScenarioTestResult {
  apiScenarioFilePath: string;
  readmeFilePath?: string;
  swaggerFilePaths: string[];
  tag?: string;

  // New added fields
  environment?: string;
  armEnv?: string;
  subscriptionId?: string;
  rootPath?: string;
  providerNamespace?: string;
  apiVersion?: string;
  startTime?: string;
  endTime?: string;
  runId?: string;
  repository?: string;
  branch?: string;
  commitHash?: string;
  baseUrl?: string;
  apiScenarioName?: string;
  stepResult: StepResult[];
}

export interface StepResult {
  statusCode: number;
  specFilePath?: string;
  exampleFilePath?: string;
  example?: SwaggerExample;
  operationId: string;
  runtimeError?: RuntimeError[];
  responseDiffResult?: ResponseDiffItem[];
  liveValidationResult?: RequestResponseLiveValidationResult;
  stepValidationResult?: any;
  correlationId?: string;
  stepName: string;
}

export interface RuntimeError {
  code: string;
  message: string;
  detail: string;
  severity: SeverityString;
}

export interface ResponseDiffItem {
  code: string;
  jsonPath: string;
  severity: SeverityString;
  message: string;
  detail: string;
}

export type ValidationLevel = "validate-request" | "validate-request-response";

export interface NewmanReportValidatorOption extends ApiScenarioLoaderOption {
  apiScenarioFilePath: string;
  reportOutputFilePath: string;
  markdown?: boolean;
  junit?: boolean;
  html?: boolean;
  specPathPrefix?: string;
  baseUrl?: string;
  runId?: string;
  validationLevel?: ValidationLevel;
  savePayload?: boolean;
  generateExample?: boolean;
  verbose?: boolean;
}

@injectable()
export class NewmanReportValidator {
  private scenario: Scenario;
  private testResult: ApiScenarioTestResult;
  private fileRoot: string;
  private liveValidator: LiveValidator;
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(
    @inject(TYPES.opts) private opts: NewmanReportValidatorOption,
    private fileLoader: FileLoader,
    private dataMasker: DataMasker,
    private swaggerAnalyzer: SwaggerAnalyzer,
    private junitReporter: JUnitReporter
  ) {
    setDefaultOpts(this.opts, {
      validationLevel: "validate-request-response",
      savePayload: false,
      generateExample: false,
      verbose: false,
    } as NewmanReportValidatorOption);
  }

  public async initialize(scenario: Scenario) {
    this.scenario = scenario;

    this.fileRoot =
      (await findReadMe(this.opts.apiScenarioFilePath)) ||
      path.dirname(this.opts.apiScenarioFilePath);

    this.testResult = {
      apiScenarioFilePath: path.relative(this.fileRoot, this.opts.apiScenarioFilePath),
      swaggerFilePaths: this.opts.swaggerFilePaths!,
      providerNamespace: getProviderFromFilePath(this.opts.apiScenarioFilePath),
      apiVersion: getApiVersionFromFilePath(this.opts.apiScenarioFilePath),
      runId: this.opts.runId,
      rootPath: this.fileRoot,
      repository: process.env.SPEC_REPOSITORY,
      branch: process.env.SPEC_BRANCH,
      commitHash: process.env.COMMIT_HASH,
      environment: process.env.ENVIRONMENT || "test",
      apiScenarioName: this.scenario.scenario,
      baseUrl: this.opts.baseUrl,
      stepResult: [],
    };

    await this.swaggerAnalyzer.initialize();

    this.liveValidator = new LiveValidator({
      fileRoot: "/",
      swaggerPaths: [...this.opts.swaggerFilePaths!],
    });
    await this.liveValidator.initialize();
  }

  public async generateReport(rawReport: NewmanReport) {
    await this.generateApiScenarioTestResult(rawReport);

    await this.outputReport();
  }

  private async generateApiScenarioTestResult(newmanReport: NewmanReport) {
    const variables = newmanReport.variables;
    this.testResult.startTime = new Date(newmanReport.timings.started).toISOString();
    this.testResult.endTime = new Date(newmanReport.timings.completed).toISOString();
    this.testResult.subscriptionId = variables.subscriptionId.value as string;
    for (const it of newmanReport.executions) {
      if (it.annotation === undefined) {
        continue;
      }
      if (it.annotation.type === "simple" || it.annotation.type === "LRO") {
        const runtimeError: RuntimeError[] = [];
        const matchedStep = this.getMatchedStep(it.annotation.step) as StepRestCall;

        // Runtime errors
        if (it.response.statusCode >= 400) {
          const error = this.getRuntimeError(it);
          runtimeError.push(error);
        }
        if (matchedStep === undefined) {
          continue;
        }

        const payload = this.convertToLiveValidationPayload(it);

        let responseDiffResult: ResponseDiffItem[] | undefined = undefined;
        const statusCode = `${it.response.statusCode}`;
        const exampleFilePath = `./examples/${matchedStep.operationId}_${statusCode}.json`;
        if (this.opts.generateExample || it.annotation.exampleName) {
          const generatedExample: SwaggerExample = {
            operationId: matchedStep.operationId,
            title: matchedStep.step,
            description: matchedStep.description,
            parameters: matchedStep._resolvedParameters!,
            responses: {
              [statusCode]: {
                headers: payload.liveResponse.headers,
                body: payload.liveResponse.body,
              },
            },
          };

          // Example validation
          if (this.opts.generateExample) {
            await this.fileLoader.writeFile(
              path.resolve(path.dirname(this.opts.reportOutputFilePath), exampleFilePath),
              JSON.stringify(generatedExample, null, 2)
            );
          }
          if (it.annotation.exampleName) {
            // validate real payload.
            responseDiffResult =
              this.opts.validationLevel === "validate-request-response"
                ? await this.exampleResponseDiff(
                    {
                      step: matchedStep.step,
                      operationId: matchedStep.operationId,
                      example: generatedExample,
                    },
                    matchedStep,
                    newmanReport
                  )
                : [];
          }
        }

        // Schema validation
        const correlationId = it.response.headers["x-ms-correlation-request-id"];
        if (this.opts.savePayload) {
          const payloadFilePath = `./payloads/${matchedStep.step}_${correlationId}.json`;
          await this.fileLoader.writeFile(
            path.resolve(path.dirname(this.opts.reportOutputFilePath), payloadFilePath),
            JSON.stringify(payload, null, 2)
          );
        }
        const liveValidationResult = await this.liveValidator.validateLiveRequestResponse(payload);

        this.testResult.stepResult.push({
          specFilePath: matchedStep.operation._path._spec._filePath,
          exampleFilePath: exampleFilePath,
          operationId: it.annotation.operationId,
          runtimeError,
          responseDiffResult: responseDiffResult,
          correlationId: correlationId,
          statusCode: it.response.statusCode,
          stepName: it.annotation.step,
          liveValidationResult: liveValidationResult,
        });
      }
    }
  }

  private convertToLiveValidationPayload(execution: NewmanExecution): RequestResponsePair {
    const request = execution.request;
    const response = execution.response;
    const liveRequest: LiveRequest = {
      url: request.url.toString(),
      method: request.method.toLowerCase(),
      headers: request.headers,
      body: request.body ? JSON.parse(request.body) : undefined,
    };
    const liveResponse: LiveResponse = {
      statusCode: response.statusCode.toString(),
      headers: response.headers,
      body: response.body ? JSON.parse(response.body) : undefined,
    };
    return {
      liveRequest,
      liveResponse,
    };
  }

  private convertPostmanFormat<T>(obj: T, convertString: (s: string) => string): T {
    if (typeof obj === "string") {
      return convertString(obj) as unknown as T;
    }
    if (typeof obj !== "object") {
      return obj;
    }
    if (obj === null || obj === undefined) {
      return obj;
    }
    if (Array.isArray(obj)) {
      return (obj as any[]).map((v) => this.convertPostmanFormat(v, convertString)) as unknown as T;
    }

    const result: any = {};
    for (const key of Object.keys(obj)) {
      result[key] = this.convertPostmanFormat((obj as any)[key], convertString);
    }
    return result;
  }

  private async exampleResponseDiff(
    example: GeneratedExample,
    matchedStep: Step,
    rawReport: NewmanReport
  ): Promise<ResponseDiffItem[]> {
    let res: ResponseDiffItem[] = [];
    if (matchedStep?.type === "restCall") {
      if (example.example.responses["200"] !== undefined) {
        Object.keys(matchedStep.variables).forEach((key) => {
          const paramName = `${matchedStep.step}_${key}`;
          matchedStep.responses = this.convertPostmanFormat(matchedStep.responses, (s) =>
            s.replace(`$(${key})`, `$(${paramName})`)
          );
        });
        res = res.concat(
          await this.responseDiff(
            example.example.responses["200"]?.body || {},
            matchedStep.responses["200"]?.body || {},
            rawReport.variables,
            `/200/body`,
            matchedStep.operation.responses["200"].schema
          )
        );
      }
    }
    return res;
  }

  private async responseDiff(
    resp: any,
    expectedResp: any,
    variables: any,
    jsonPathPrefix: string,
    schema: any
  ): Promise<ResponseDiffItem[]> {
    const env = new VariableEnv(variables);
    try {
      const ignoredKeys = [/\/body\/id/, /\/body\/location/, /date/i];
      const ignoreValuePattern = [/\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d(?:\.\d+)?Z?/];
      const expected = env.resolveObjectValues(expectedResp);
      const delta: ResponseDiffItem[] = getJsonPatchDiff(expected, resp, {
        includeOldValue: true,
        minimizeDiff: false,
      })
        .map((it: any) => {
          const ret: ResponseDiffItem = {
            code: "",
            jsonPath: "",
            severity: "Error",
            message: "",
            detail: "",
          };
          const jsonPath: string = it.remove || it.add || it.replace;
          const propertySchema = this.swaggerAnalyzer.findSchemaByJsonPointer(
            jsonPath,
            schema,
            resp
          );
          if (propertySchema["x-ms-secret"] === true) {
            return undefined;
          }
          if (it.remove !== undefined) {
            ret.code = "RESPONSE_MISSING_VALUE";
            ret.jsonPath = jsonPathPrefix + it.remove;
            ret.severity = "Error";
            ret.message = `The response value is missing. Path: ${
              ret.jsonPath
            }. Expected: ${this.dataMasker.jsonStringify(it.oldValue)}. Actual: undefined`;
          } else if (it.add !== undefined) {
            ret.code = "RESPONSE_ADDITIONAL_VALUE";
            ret.jsonPath = jsonPathPrefix + it.add;
            ret.severity = "Error";
            ret.message = `Return additional response value. Path: ${
              ret.jsonPath
            }. Expected: undefined. Actual: ${this.dataMasker.jsonStringify(it.value)}`;
          } else if (it.replace !== undefined) {
            ret.code = "RESPONSE_INCONSISTENT_VALUE";
            ret.jsonPath = jsonPathPrefix + it.replace;

            const paths = it.replace.split("/");
            while (paths.length > 1) {
              const path = paths.join("/");
              const sch = this.swaggerAnalyzer.findSchemaByJsonPointer(path, schema, resp);
              // Ignore diff when propertySchema readonly is true. When property is readonly, it's probably a random generated value which updated dynamically per request.
              if (sch.readOnly === true) {
                return undefined;
              }
              paths.pop();
            }
            ret.severity = "Error";
            ret.message = `The actual response value is different from example. Path: ${
              ret.jsonPath
            }. Expected: ${this.dataMasker.jsonStringify(
              it.oldValue
            )}. Actual: ${this.dataMasker.jsonStringify(it.value)}`;
          }
          ret.detail = this.dataMasker.jsonStringify(it);
          if (
            ignoredKeys.some((key) => ret.jsonPath.search(key) !== -1) ||
            (ret.code === "RESPONSE_INCONSISTENT_VALUE" &&
              typeof it.value === "string" &&
              ignoreValuePattern.some((valuePattern) => it.value.search(valuePattern) !== -1))
          ) {
            return undefined;
          }
          return ret;
        })
        .filter((it) => it !== undefined) as ResponseDiffItem[];
      return delta;
      // eslint-disable-next-line no-empty
    } catch (err) {}
    return [];
  }

  private getMatchedStep(stepName: string): Step | undefined {
    for (const it of this.scenario.steps ?? []) {
      if (stepName === it.step) {
        return it;
      }
    }
    return undefined;
  }

  private getRuntimeError(it: NewmanExecution): RuntimeError {
    const responseObj = this.dataMasker.jsonParse(it.response.body);
    return {
      code: "RUNTIME_ERROR",
      message: `statusCode: ${it.response.statusCode}, errorCode: ${responseObj?.error?.code}, errorMessage: ${responseObj?.error?.message}`,
      severity: "Error",
      detail: this.dataMasker.jsonStringify(it.response.body),
    };
  }

  private async outputReport(): Promise<void> {
    if (this.opts.reportOutputFilePath !== undefined) {
      console.log(`Write generated report file: ${this.opts.reportOutputFilePath}`);
      await this.fileLoader.writeFile(
        this.opts.reportOutputFilePath,
        JSON.stringify(this.testResult, null, 2)
      );
    }
    if (this.opts.markdown) {
      await this.fileLoader.appendFile(
        path.join(path.dirname(path.dirname(this.opts.reportOutputFilePath)), "report.md"),
        generateMarkdownReport(this.testResult)
      );
    }
    if (this.opts.junit) {
      await this.junitReporter.addSuiteToBuild(
        this.testResult,
        path.join(path.dirname(path.dirname(this.opts.reportOutputFilePath)), "junit.xml")
      );
    }
  }
}
