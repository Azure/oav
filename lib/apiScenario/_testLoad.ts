import "reflect-metadata";

import { dirname } from "path";
import { getDefaultAzureCredential } from "@azure/identity";
import { ApiScenarioLoader } from "./apiScenarioLoader";
import { ApiScenarioRunner } from "./apiScenarioRunner";
import { VariableEnv } from "./variableEnv";
import { ApiScenarioRestClient } from "./apiScenarioRestClient";

const main = async () => {
  const readmeMd: string =
    "/home/htc/azure-rest-api-specs/specification/containerservice/resource-manager/readme.md";
  // "/home/htc/azure-rest-api-specs/specification/network/resource-manager/readme.md";
  // "/home/htc/azure-rest-api-specs/specification/operationalinsights/resource-manager/readme.md";
  const fileRoot = dirname(readmeMd);

  const loader = ApiScenarioLoader.create({
    useJsonParser: false,
    checkUnderFileRoot: false,
    fileRoot,
  });

  const testDef = await loader.load(
    // "Microsoft.ContainerService/stable/2020-07-01/test-scenarios/testAks.yml"
    // "Microsoft.OperationalInsights/stable/2020-08-01/test-scenarios/testDataExport.yaml"
    // "Microsoft.Network/stable/2020-08-01/test-scenarios/testNetworkPublicIp.yaml"
    "Microsoft.ContainerService/stable/2020-12-01/test-scenarios/containerService.yaml"
  );

  console.log(testDef.scenarios[0].steps);

  const env = new VariableEnv();
  env.setBatchEnv({
    subscriptionId: "db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    location: "westus",
    SSH_PUBLIC_KEY: "__public_key_ssh__",
  });

  const runner = new ApiScenarioRunner({
    jsonLoader: loader.jsonLoader,
    env,
    client: new ApiScenarioRestClient(getDefaultAzureCredential(), {}),
  });

  try {
    // for (const scenario of testDef.testScenarios) {
    //   await runner.executeScenario(scenario);
    // }
    await runner.executeScenario(testDef.scenarios[0]);
  } catch (e) {
    console.log(e);
  } finally {
    console.timeLog("TestLoad");
    await runner.cleanAllScope();
  }
};

console.time("TestLoad");
console.log("Start");

main()
  .catch((e) => {
    console.error(e);
  })
  .finally(() => {
    console.timeEnd("TestLoad");
  });
