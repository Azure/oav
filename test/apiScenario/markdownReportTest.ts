// Copyright (c) 2021 Microsoft Corporation
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
import {
  generateMarkdownReport,
  generateMarkdownReportHeader,
} from "../../lib/apiScenario/markdownReport";
import { ApiScenarioTestResult } from "../../lib/apiScenario/newmanReportValidator";

describe("markdownReport", () => {
  it("Should generate markdown report", () => {
    const ts = {
      apiScenarioFilePath: "Microsoft.Compute/preview/2020-09-30/test-scenarios/galleries.yaml",
      swaggerFilePaths: ["Microsoft.Compute/preview/2020-09-30/gallery.json"],
      providerNamespace: "Microsoft.Compute",
      apiVersion: "2020-09-30",
      runId: "202106011456-d4udg",
      rootPath: "/home/zhenglai/repos/azure-rest-api-specs/specification/compute/resource-manager",
      environment: "test",
      apiScenarioName: "galleries_1",
      armEndpoint: "https://management.azure.com",
      stepResult: [
        {
          exampleFilePath:
            "Microsoft.Compute/preview/2020-09-30/test-scenarios/examples/CreateOrUpdateASimpleGalleryWithSharingProfile.json",
          operationId: "Galleries_CreateOrUpdate",
          runtimeError: [
            {
              code: "RUNTIME_ERROR_400",
              message:
                "code: BadRequest, message: Subscription 'db5eb68e-73e2-4fa8-b18a-46cd1be4cce5' is not registered with the feature Microsoft.Compute/SIGSharing. Please register the subscription before retrying the call.",
              severity: "Error",
              detail:
                '{\r\n  "error": {\r\n    "code": "BadRequest",\r\n    "message": "Subscription \'db5eb68e-73e2-4fa8-b18a-46cd1be4cce5\' is not registered with the feature Microsoft.Compute/SIGSharing. Please register the subscription before retrying the call."\r\n  }\r\n}',
            },
          ],
          statusCode: 400,
          stepName: "Create or update a simple gallery with sharing profile.",
        },
        {
          exampleFilePath:
            "Microsoft.Compute/preview/2020-09-30/test-scenarios/examples/DeleteAGallery.json",
          operationId: "Galleries_Delete",
          runtimeError: [],
          statusCode: 204,
          stepName: "Delete a gallery.",
        },
      ],
      startTime: "2021-06-01T06:56:46.835Z",
      endTime: "2021-06-01T06:58:07.066Z",
      subscriptionId: "db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    } as ApiScenarioTestResult;
    const header = generateMarkdownReportHeader();
    const body = generateMarkdownReport(ts);
    expect(header).toMatchSnapshot("header");
    expect(body).toMatchSnapshot("body");
  });
});
