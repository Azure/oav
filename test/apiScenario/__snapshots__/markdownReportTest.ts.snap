// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdownReport Should generate markdown report: body 1`] = `
"
  <details>
    <summary>
      <code>
        ❌
      </code>
      <b>
        galleries_1: 1 Fatal Step(s)
      </b>
    </summary>
    <br />
    <ul>
      <li>
        Duration:
        1 minute, 20.231 seconds
      </li>
      <li>
        Swagger Files
        <ul>
            <li>
              <a target=\\"_blank\\" href=\\"Microsoft.Compute/preview/2020-09-30/gallery.json\\">
                Microsoft.Compute/preview/2020-09-30/gallery.json
              </a>
            </li>
        </ul>
      </li>
      <li>
        Run Id: 202106011456-d4udg
      </li>
    </ul>
    <b>
      &nbsp;&nbsp;&nbsp;&nbsp;API Test Steps:
    </b>
    <ol>
<li>
  <code>
        ❌
  </code>
<b>
        Create or update a simple gallery with sharing profile.: 1 Fatal Error(s)
</b>
</li>
<ul>
    <li>
      StatusCode:
      400
    </li>
    <li>
      Duration:
      0 seconds
    </li>
  </ul>

|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Rule&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Message |
|------|---------|
| <code> ❌ </code> [RUNTIME_ERROR_400](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/api-scenario/references/ErrorCodeReference.md#RUNTIME_ERROR_400) | code: BadRequest, message: Subscription 'db5eb68e-73e2-4fa8-b18a-46cd1be4cce5' is not registered with the feature Microsoft.Compute/SIGSharing. Please register the subscription before retrying the call. |
<br />
<li>
  <code>
        ️✔️
  </code>
<b>
        Delete a gallery.
</b>
</li>
<ul>
    <li>
      StatusCode:
      204
    </li>
    <li>
      Duration:
      0 seconds
    </li>
  </ul>

<br />

</ol>
</details>
"
`;

exports[`markdownReport Should generate markdown report: header 1`] = `"<h3>Azure API Test Report</h3>"`;
